src/psyclone/psyir/nodes/operation.py

```diff
@@ -224,38 +230,137 @@ def create(operator, lhs, rhs):
+                            dtype.intrinsic.partial_datatype):
+                        dtype = ArrayType(dtype.intrinsic.partial_datatype,
+                                          shape=dtype.shape)
+                    else:
+                        return UnknownFortranType("")
+
+            argtypes.append(dtype)
+
+        base_type = None
+
+        if self.operator not in self._numeric_ops:
+            # Must be a relational or logical operator. Intrinsic type of
+            # result will be boolean.
+            base_type = BOOLEAN_TYPE
+        else:
+            # We have a numerical operation.
+            if argtypes[0].intrinsic == argtypes[1].intrinsic:
+                if argtypes[0].precision != argtypes[1].precision:
+                    raise NotImplementedError(
+                        f"Cannot determine the type of expression "
+                        f"'{self.debug_string()}' involving arguments of the "
+                        f"same intrinsic type but different precision "
+                        f"('{argtypes[0].precision}' and "
+                        f"'{argtypes[1].precision}').")
+                # Operands are of the same type so that is the type of the
+                # result.
+                base_type = argtypes[0]
+            elif argtypes[0].intrinsic == ScalarType.Intrinsic.REAL:
+                base_type = argtypes[0]
+            elif argtypes[1].intrinsic == ScalarType.Intrinsic.REAL:
+                base_type = argtypes[1]
+
+            # Check that the type of the result is consistent with a numerical
+            # operation.
+            if not base_type or base_type.intrinsic not in (
+                    ScalarType.Intrinsic.INTEGER,
+                    ScalarType.Intrinsic.REAL):
+                for atype in argtypes:
+                    if atype.intrinsic not in (ScalarType.Intrinsic.INTEGER,
+                                               ScalarType.Intrinsic.REAL):
+                        raise TypeError(
+                            f"Invalid argument of type '{atype.intrinsic}' to "
+                            f"numerical operation '{self.operator}' in "
+                            f"'{self.debug_string()}'")
+
+        if all(isinstance(atype, ScalarType) for atype in argtypes):
+            # Both operands are of scalar type.
+            return base_type
+
+        if all(isinstance(atype, ArrayType) for atype in argtypes):
+            # Both operands are of array type.
+            if len(argtypes[0].shape) != len(argtypes[1].shape):
+                raise TypeError(
+                    f"Binary operation '{self.debug_string()}' has operands "
+                    f"of different shape: '{self.children[0].debug_string()}' "
+                    f"has rank {len(argtypes[0].shape)} and "
+                    f"'{self.children[1].debug_string()}' has rank "
+                    f"{len(argtypes[1].shape)}")
```

Comment:
Shouldn't this return the larger shape (with the same assumption that extends match)? e.g:
```
real, dimension(:,:) :: A
real, dimension(:) :: B

A + B
```
result is equal to type of A

Reply:
This isn't supported because it is ambiguous. e.g. gfortran says:

        6 |   write(*,*) A + B
          |            1     2
    Error: Inconsistent ranks for operator at (1) and (2)